GViewCore/src/View/DissasmViewer/DissasmFunctionUtils.cpp

```diff
@@ -0,0 +1,228 @@
+    }
+
+    if (!size || !start)
+        return false;
+
+    if (size < 2) {
+        ptr = !is_hex ? op_str : op_str + 2;
+        while (ptr && *ptr != '\0') {
+            if (!(*ptr >= '0' && *ptr <= '9' || *ptr >= 'a' && *ptr <= 'f'))
+                return false;
+            ptr++;
+        }
+    }
+
+    if (!checkValidSequence())
+        return false;
+
+    const NumberParseFlags numberFlags = is_hex ? NumberParseFlags::Base16 : NumberParseFlags::Base10;
+    const auto sv                      = std::string_view(start, size);
+    const auto converted               = Number::ToUInt64(sv, numberFlags);
+    if (!converted.has_value())
+        return false;
+
+    value = converted.value();
+
+    return true;
+}
+
+LocalString<64> FormatFunctionName(uint64 functionAddress, const char* prefix)
+{
+    NumericFormatter formatter;
+    auto sv = formatter.ToHex(functionAddress);
+    LocalString<64> callName;
+    callName.AddFormat("%s%09s", prefix, sv.data());
+    return callName;
+}
+
+AsmOffsetLine SearchForClosestAsmOffsetLineByOffset(const std::vector<AsmOffsetLine>& values, uint64 searchedOffset)
+{
+    assert(!values.empty());
+    uint32 left  = 0;
+    uint32 right = static_cast<uint32>(values.size()) - 1u;
+    while (left < right) {
+        const uint32 mid = (left + right) / 2;
+        if (searchedOffset == values[mid].offset)
+            return values[mid];
+        if (searchedOffset < values[mid].offset)
+            right = mid - 1;
+        else
+            left = mid + 1;
+    }
+    if (left > 0 && values[left].offset > searchedOffset)
+        return values[left - 1];
+
+    return values[left];
+}
+
+cs_insn* GetCurrentInstructionByOffset(
+      uint64 offsetToReach, DissasmCodeZone* zone, Reference<GView::Object> obj, uint32& diffLines, DrawLineInfo* dli)
+{
+    const auto closestData = SearchForClosestAsmOffsetLineByOffset(zone->cachedCodeOffsets, offsetToReach);
+    zone->lastClosestLine  = closestData.line;
+    zone->asmAddress       = closestData.offset - zone->cachedCodeOffsets[0].offset;
+    zone->asmSize          = zone->zoneDetails.size - zone->asmAddress;
+
+    // TODO: maybe get less data ?
+    const auto instructionData = obj->GetData().Get(zone->cachedCodeOffsets[0].offset + zone->asmAddress, static_cast<uint32>(zone->asmSize), false);
+    zone->lastData             = instructionData;
+    if (!instructionData.IsValid()) {
+        if (dli)
+            dli->WriteErrorToScreen("ERROR: extract valid data from file!");
+        diffLines = UINT32_MAX;
+        return nullptr;
+    }
+
+    zone->asmData = const_cast<uint8*>(zone->lastData.GetData());
+
+    // TODO: keep the handle open and insn open until the program ends
+    csh handle;
+    const auto resCode = cs_open(CS_ARCH_X86, static_cast<cs_mode>(zone->internalArchitecture), &handle);
+    if (resCode != CS_ERR_OK) {
+        if (dli)
+            dli->WriteErrorToScreen(cs_strerror(resCode));
+        cs_close(&handle);
```

Comment:
defer

Reply:
Will be done in [this](https://github.com/gdt050579/GView/issues/330)